File: python/dsa/data-structures/graph/cycle.py
def hasCycle(u, p, visited, adj): # @viz:def
    visited.add(u) # @viz:mark
    
    for v in adj[u]: # @viz:for
        if v == p: # @viz:skip_check
            continue # @viz:skip_body
        
        if v in visited: # @viz:cycle_check
            return True # @viz:cycle_body
            
        if hasCycle(v, u, visited, adj): # @viz:dfs_check @viz:dfs_body
            return True # @viz:dfs_ret
            
    return False

File: java/dsa/data-structures/graph/cycle.java
boolean hasCycle(int u, int p, boolean[] visited, List<List<Integer>> adj) { // @viz:def
    visited[u] = true; // @viz:mark
    
    for (int v : adj.get(u)) { // @viz:for
        if (v == p) { // @viz:skip_check
            continue; // @viz:skip_body
        }
        
        if (visited[v]) { // @viz:cycle_check
            return true; // @viz:cycle_body
        }
        
        if (hasCycle(v, u, visited, adj)) { // @viz:dfs_check @viz:dfs_body
            return true; // @viz:dfs_ret
        }
    }
    return false;
}